src/Cooked/MockChain/GenerateTx/Collateral.hs

```diff
@@ -44,59 +43,62 @@ toCollateralTriplet ::
+        return $ mconcat (Api.txOutValue <$> collateralInsResolved)
+      -- We retrieve the collateral percentage compared to fees. By default, we use
+      -- 150% which is the current value in the parameters, although the default
+      -- value should never be used here, as the call is supposed to always succeed.
+      collateralPercentage <- asks (toInteger . fromMaybe 150 . Cardano.protocolParamCollateralPercent . Emulator.pProtocolParams . params)
+      -- The total collateral corresponds to the fees multiplied by the collateral
+      -- percentage. We add 1 because the ledger apparently rounds up this value.
+      coinTotalCollateral <- asks (Emulator.Coin . (+ 1) . (`div` 100) . (* collateralPercentage) . fee)
+      -- We create the total collateral based on the computed value
+      let txTotalCollateral = Cardano.TxTotalCollateral Cardano.BabbageEraOnwardsConway coinTotalCollateral
+      -- We compute a return collateral value by subtracting the total collateral to
+      -- the value in collateral inputs
+      let returnCollateralValue = collateralInsValue <> PlutusTx.negate (toValue coinTotalCollateral)
+      -- This should never happen, as we always compute the collaterals for the
+      -- user, but we guard against having some negative elements in the value in
+      -- case we give more freedom to the users in the future
+      when (fst (Api.split returnCollateralValue) /= mempty) $ throwOnString "toCollateralTriplet: negative parts in return collateral value"
+      -- The return collateral is then computed
+      txReturnCollateral <-
+        -- If the total collateral equal what the inputs provide, we return
+        -- `TxReturnCollateralNone`, otherwise, we compute the new output
+        if returnCollateralValue == mempty
+          then return Cardano.TxReturnCollateralNone
+          else do
+            -- The value is a translation of the remaining value
+            txReturnCollateralValue <-
+              Ledger.toCardanoTxOutValue
+                <$> throwOnToCardanoError
+                  "toCollateralTriplet: cannot build return collateral value"
+                  (Ledger.toCardanoValue returnCollateralValue)
+            -- The address is the one from the return collateral wallet, which is
+            -- required to exist here.
+            address <- do
+              networkId <- asks (Emulator.pNetworkId . params)
+              throwOnToCardanoError "toCollateralTriplet: cannot build return collateral address" $
+                Ledger.toCardanoAddressInEra networkId (walletAddress returnCollateralWallet)
+            -- The return collateral is built up from those elements
+            return $
+              Cardano.TxReturnCollateral Cardano.BabbageEraOnwardsConway $
+                Cardano.TxOut address txReturnCollateralValue Cardano.TxOutDatumNone Cardano.ReferenceScriptNone
+      return (txInsCollateral, txTotalCollateral, txReturnCollateral)
```

Comment:
I see a big chunk of diff here. I did not check in details. Is it just reformatting / simple adaptations.